File: apps/ios/Pulse/Features/Calls/CallVideoViews.swift
import SwiftUI
import WebRTC

// ─────────────────────────────────────────────────────────────
// Pulse — Wave R1-W2D REAL video renderers for the call overlay (web
// call-overlay.tsx parity):
//   • remote video → full-bleed MTK surface (web `absolute inset-0 cover`)
//   • local camera → mirrored PiP tile (web `scaleX(-1)`, ≈96×144pt rounded
//     bordered tile above the controls)
//
// Both wrap RTCMTLVideoView (Metal, the modern default) bound to a REAL
// RTCVideoTrack via add/remove sink. Renderer lifecycle: the sink detaches
// and the view releases with SwiftUI's dismantle — no camera surface leaks.
// ─────────────────────────────────────────────────────────────

/// One RTCVideoTrack rendered through RTCMTLVideoView. Recreated per track
/// instance; mirrored self-views flip the UIKit transform (web scaleX(-1)).
struct VideoTrackView: UIViewRepresentable {
    let track: RTCVideoTrack
    var mirrored: Bool = false

    final class Coordinator {
        let track: RTCVideoTrack
        init(track: RTCVideoTrack) { self.track = track }
    }

    func makeCoordinator() -> Coordinator {
        Coordinator(track: track)
    }

    func makeUIView(context: Context) -> RTCMTLVideoView {
        let view = RTCMTLVideoView(frame: .zero)
        // Web object-cover parity — fill, crop, keep aspect.
        view.videoContentMode = .scaleAspectFill
        if mirrored {
            view.transform = CGAffineTransform(scaleX: -1, y: 1)
        }
        context.coordinator.track.add(view)
        return view
    }

    func updateUIView(_ uiView: RTCMTLVideoView, context: Context) {}

    static func dismantleUIView(_ uiView: RTCMTLVideoView, coordinator: Coordinator) {
        coordinator.track.remove(uiView)
    }
}

/// The peer's video, full-bleed behind the call UI (web remote <video>).
/// Renders only once the remote track actually arrives — the audio-only
/// fallback keeps the avatar-only layout.
struct RemoteVideoLayer: View {
    @ObservedObject var engine: PulseCallEngine

    var body: some View {
        ZStack {
            if let track = engine.remoteVideoTrack {
                VideoTrackView(track: track, mirrored: false)
                    .background(Color.black)
                    .ignoresSafeArea()
            }
        }
    }
}

/// The local camera PiP — mirrored self-view tile, bottom-trailing above the
/// controls (web `bottom-28 right-4` parity). Hidden while the camera toggle
/// is off (web cameraEnabled parity) or when no camera is attached.
struct LocalVideoPipLayer: View {
    @ObservedObject var engine: PulseCallEngine

    var body: some View {
        Group {
            if engine.cameraEnabled, let track = engine.localVideoTrack {
                VideoTrackView(track: track, mirrored: true)
                    .frame(width: 96, height: 144)
                    .clipShape(RoundedRectangle(cornerRadius: 18, style: .continuous))
                    .overlay(
                        RoundedRectangle(cornerRadius: 18, style: .continuous)
                            .strokeBorder(Color.white.opacity(0.15), lineWidth: 1)
                    )
                    .shadow(color: .black.opacity(0.4), radius: 10, y: 5)
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .bottomTrailing)
        .padding(.trailing, 16)
        .padding(.bottom, 168)
        .allowsHitTesting(false)
    }
}

File: apps/android/feature-calls/src/main/kotlin/app/pulse/feature/calls/CallVideoViews.kt
package app.pulse.feature.calls

import androidx.compose.foundation.border
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.compose.ui.viewinterop.AndroidView
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import org.webrtc.EglBase
import org.webrtc.SurfaceViewRenderer
import org.webrtc.VideoTrack

/**
 * Wave R1-W2D — REAL WebRTC video renderers for the call overlay (web
 * call-overlay.tsx parity):
 *   • remote video  → full-bleed surface (web `absolute inset-0 object-cover`)
 *   • local camera  → small PiP self-view, mirrored (web `scaleX(-1)`, 96×144dp
 *     ≈ the web h-36 w-24 tile, rounded border)
 *
 * Both wrap [SurfaceViewRenderer] and MUST share the engine's factory EGL
 * context. Sink attach/detach + renderer release are lifecycle-driven so
 * camera surfaces never leak past the call.
 */

/**
 * One video track on one renderer. Recreated per track instance — addSink
 * happens after the effect commits, release on dispose (no double-sink).
 */
@Composable
fun VideoRenderer(
    track: VideoTrack?,
    eglBaseContext: EglBase.Context,
    mirrored: Boolean,
    modifier: Modifier = Modifier,
) {
    if (track == null) return
    val context = androidx.compose.ui.platform.LocalContext.current
    val renderer = remember(track) { SurfaceViewRenderer(context) }
    DisposableEffect(track) {
        renderer.init(eglBaseContext, null)
        if (mirrored) renderer.setMirror(true)
        runCatching { track.addSink(renderer) }
        onDispose {
            runCatching { track.removeSink(renderer) }
            runCatching { renderer.release() }
        }
    }
    AndroidView(
        factory = { renderer },
        modifier = modifier,
    )
}

/**
 * The peer's video, full-bleed behind the call UI. Renders only once the
 * remote video track actually arrives over the negotiated m=video line.
 */
@Composable
fun RemoteVideoSurface(engine: CallEngine, modifier: Modifier = Modifier) {
    val track by engine.remoteVideoTrack.collectAsStateWithLifecycle()
    VideoRenderer(
        track = track,
        eglBaseContext = engine.eglBaseContext,
        mirrored = false,
        modifier = modifier,
    )
}

/**
 * The local camera PiP — mirrored self-view (web parity), bottom-end tile
 * above the controls. Only rendered while capture is actually running AND
 * the camera toggle is on (muted camera ⇒ hidden, web cameraEnabled parity).
 */
@Composable
fun LocalVideoPip(engine: CallEngine, modifier: Modifier = Modifier) {
    val track by engine.localVideoTrack.collectAsStateWithLifecycle()
    val cameraEnabled by engine.cameraEnabled.collectAsStateWithLifecycle()
    if (!cameraEnabled) return
    VideoRenderer(
        track = track,
        eglBaseContext = engine.eglBaseContext,
        mirrored = true,
        modifier = modifier
            .width(96.dp)
            .height(144.dp)
            .clip(RoundedCornerShape(16.dp))
            .border(1.dp, Color.White.copy(alpha = 0.15f), RoundedCornerShape(16.dp)),
    )
}
